pkg/cvo/metrics.go

```diff
@@ -105,38 +115,84 @@ type asyncResult struct {
 	error error
 }
 
-// RunMetrics launches an server bound to listenAddress serving
-// Prometheus metrics at /metrics over HTTPS.  Continues serving
-// until runContext.Done() and then attempts a clean shutdown
-// limited by shutdownContext.Done().  Assumes runContext.Done()
-// occurs before or simultaneously with shutdownContext.Done().
-func RunMetrics(runContext context.Context, shutdownContext context.Context, listenAddress string, tlsConfig *tls.Config) error {
-	if tlsConfig == nil {
-		return errors.New("TLS configuration is required to serve metrics")
-	}
-
+func createHttpServer() *http.Server {
 	handler := http.NewServeMux()
 	handler.Handle("/metrics", promhttp.Handler())
 	server := &http.Server{
 		Handler: handler,
 	}
+	return server
+}
 
-	tcpListener, err := net.Listen("tcp", listenAddress)
+func shutdownHttpServer(parentCtx context.Context, svr *http.Server) {
+	ctx, cancel := context.WithTimeout(parentCtx, 5*time.Second)
+	defer cancel()
+	klog.Info("Shutting down metrics server so it can be recreated with updated TLS configuration.")
+	if err := svr.Shutdown(ctx); err != nil {
+		klog.Errorf("Failed to gracefully shut down metrics server during restart: %w", err)
```

Comment:
I think `%w` should be for `fmt.Errorf`, but not for `klog.Errorf`.  Checking locally:

```console
$ cat test.go 
package main

import (
        "fmt"
        "k8s.io/klog/v2"
)

func main() {
        err := fmt.Errorf("a")
        err = fmt.Errorf("b: %w", err)
        klog.Errorf("w: %w", err)
        klog.Errorf("v: %v", err)
}
$ go run test.go 
E0114 15:13:06.218632    8539 test.go:11] w: %!w(*fmt.wrapError=&{b: a 0xc000098380})
E0114 15:13:06.218733    8539 test.go:12] v: b: a
```

So `%w` rendering strings, `fmt.Errorf` is identical to `%v`, but `klog.Errorf` is a noisy Go dump business.

Reply:
There are a few more `klog.Errorf("...%w...", ...)` calls in the current tip that I'd prefer used `%v`:

```console
$ git show origin/pr/726 | grep '^+.*klog.Errorf.*%w'
+               klog.Errorf("Failed to gracefully shut down metrics server during restart: %w", err)
+                               klog.Errorf("Failed to gracefully shut down metrics server: %w", shutdownError)
+               klog.Errorf("Failed to open file %v for checksum: %w", fName, err)
+               klog.Errorf("Failed to compute checksum for file %v: %w", fName, err)
```

Reply:
As of 0e6136c727b05509c:

```console
$ git show origin/pr/726 | grep -c '^+.*klog.Errorf.*%w'
0
$ git show origin/pr/726 | grep '^+.*fmt.Errorf' | grep -v '%w'
+               return false, fmt.Errorf("Certificate or key is missing or empty, certificates will not be rotated.")
```

So none of the `klog.Errorf` calls use `%w` (which is good).  And all of the `fmt.Errrof` calls use `%w` (which is good), except for a single entry that's not actually formatting strings at all.  That could use [`errors.New`][1] instead, but isn't a big deal, and I'm going to mark this thread resolved without waiting on that last nit.

[1]: https://pkg.go.dev/errors#New